### ISSUE TITLE
Add a property to disable HTTP Observations for actuator endpoints

### ISSUE THREAD

**Issue by jonatan-ivanov** [spring-projects/spring-boot#34801]
It seems there is a good amount of users who are registering an `ObservationPredicate` in order to disable Observations for `/actuator/**` endpoints, something like this:

```java
@Bean
ObservationPredicate actuatorServerContextPredicate() {
    return (name, context) -> {
        if (name.equals("http.server.requests") && context instanceof ServerRequestObservationContext serverContext) {
            return !serverContext.getCarrier().getRequestURI().startsWith("<actuator-base-path>");
        }
        return true;
    };
}
```

I think this common use-case could be simplified by creating a similar bean (for mvc and webflux) and let the users to configure this using a single property.

**Comment by jonatan-ivanov** [comment]
Connected issue in Framework: https://github.com/spring-projects/spring-framework/issues/29210

**Comment by Saljack** [comment]
I implemented this solution in our services and it makes things worse than better because there is also Spring Security Observation. The security observation is not affected (because) this predicate disables only `http.server.requests`.  This also creates multiple separated traces because they are created out of any parent trace id (`http.server.requests` trace). So then there are for example these traces:
```
643d472f5d60ba294b18360a79427310 - api-gateway: security filterchain after
643d472b0d5bde11b0fc3b53edc9ed36 - api-gateway: authorize -security-context-server-web-exchange
643d472646d2ef6b55c842b12e41d502 - api-gateway: security filterchain before
```
Of course I can disable Spring Security Observation but it does not make sense because then there can be for example Spring Data Observation.

**Comment by jonatan-ivanov** [comment]
There are a few connected issues to this, let me try to collect them here:
- https://github.com/micrometer-metrics/micrometer/issues/3678
- https://github.com/spring-projects/spring-security/issues/12854
- https://github.com/spring-projects/spring-framework/issues/29210
- https://github.com/spring-projects/spring-boot/issues/34400

What you described as an issue can be a valid use-case for other users (ignoring parent observation but keeping children) but there is a solution/workaround to disable every Observation that was triggered through a specific HTTP endpoint (e.g.: `/actuator`) in https://github.com/spring-projects/spring-security/issues/12854, please see https://github.com/spring-projects/spring-security/issues/12854#issuecomment-1490838037 and https://github.com/spring-projects/spring-security/issues/12854#issuecomment-1491992099

**Comment by Saljack** [comment]
Ohh thanks. It is a nice summary. I was totally blind 🤦. Now it makes more sense to me. It is pretty hard to follow everything regarding observation because you need to follow/search multiple projects. I believe that most of these issues will be resolved in the Spring Boot 3.1. I really appreciate your work 👍.

**Comment by jonatan-ivanov** [comment]
I might have something that could work for you but I have never tried with webflux and I did not test it a lot for webmvc, I threw it out because of the `RequestContextHolder` solution above seemed much better:

I think you might be able to write a `getRoot` method that recursively can go upwards on the parent chain (`context.parent`) and find the one that has no (`null`) parent. At that point, you get the root context/observation, something like this:

```java
private Observation.Context getRoot(Observation.Context current) {
    ObservationView parent = current.getParentObservation();
    if (parent == null) {
        return current;
    }
    else {
        return getRoot((Context) parent.getContextView());
    }
}
```
And if you got the root, you can check if it is an "HTTP context" and the path so you can ignore actuator, something like this:
```java
@Bean
ObservationPredicate noActuatorServerObservations() {
    return (name, context) ->
        Context root = getRoot(context);
        if (context instanceof ServerRequestObservationContext serverContext) {
            return !serverContext.getCarrier().getRequestURI().startsWith("/actuator");
        }
        else {
            return true;
        }
    };
}
```

I guess another workaround could be getting the current request from Reactor somehow, maybe @chemicL or @bclozel could help in that.

**Comment by chemicL** [comment]
From Reactor's perspective, I think not much can be done, aside from clearing the [ObservationThreadLocalAccessor.KEY](https://github.com/micrometer-metrics/micrometer/blob/f6f8eaed9fe7ff1399964bf85b7b850ad0732187/micrometer-observation/src/main/java/io/micrometer/observation/contextpropagation/ObservationThreadLocalAccessor.java#L33) from the Reactor `Context`, but that doesn't disable the `Observation` that is downstream in the operator chain: if it's started, it is in effect, only the propagation would not reach some parts, but the framework codebase is where the creation and start process needs to be eliminated.

**Comment by bclozel** [comment]
@saulgiordani can you create a new issue in the Spring Framework project with a sample application reproducing the problem? Thanks!

**Comment by saulgiordani** [comment]
> @saulgiordani can you create a new issue in the Spring Framework project with a sample application reproducing the problem? Thanks!

Hi @bclozel, the issue was on my side. I implemented the interface ServerRequestObservationConvention without overriding the getName() method, this is why it was returning null

**Comment by goafabric** [comment]
@gofabric: remindme

**Comment by codefromthecrypt** [comment]
if it helps progress this, I would suggest peeling off the health endpoint as something easier to land maybe? Use in docker and k8s fills up trace repositories with traces named like `http get /actuator/health/{*path}`

This is a bad experience for those migrating from sleuth or anything else that skips health by default

**Comment by jonatan-ivanov** [comment]
@codefromthecrypt You might be able to do this as a workaround:
```java
@Bean
ObservationPredicate noActuatorServerObservations() {
    return (name, context) -> {
        if (name.equals("http.server.requests") && context instanceof ServerRequestObservationContext serverContext) {
            return !serverContext.getCarrier().getRequestURI().startsWith("/actuator");
        }
        else {
            return true;
        }
    };
}
```

**Comment by codefromthecrypt** [comment]
note @nidhi-nair seemed to need to look up a root context to implement this, though I don't know if it was project-specific https://github.com/appsmithorg/appsmith/pull/24758

**Comment by codefromthecrypt** [comment]
@jonatan-ivanov thanks I was able to get something similar working https://github.com/openzipkin/brave-example/pull/120

**Comment by codefromthecrypt** [comment]
one thing that became apparent in this exercise is that this is the perfect thing to normalize in a property at least until there's a generic HTTP path accessor. Depending on if you are reactive or not, you have to define this filter with different types. I think that's not ideal especially for something that used to work via property.

**Comment by jonatan-ivanov** [comment]
>note @nidhi-nair seemed to need to look up a root context to implement this, though I don't know if it was project-specific

Yes, you need to find the root in some use-cases. I would say it is use-case specific: if you have more than one spans and you don't have access to the original request (or something that the root has), you need to look the root up, here's another use case with some possible shortcuts in Spring Security: https://github.com/spring-projects/spring-security/issues/12854

>@jonatan-ivanov thanks I was able to get something similar working https://github.com/openzipkin/brave-example/pull/120

👍🏼

>one thing that became apparent in this exercise is that this is the perfect thing to normalize in a property [...]

I agree, I also think that this could be more generic (not actuator-only) where the users are able to define the path prefix.

**Comment by mfvanek** [comment]
We run our Spring Boot apps on Kubernetes.
We always move actuator to the separate port.
We collect liveness probe from actuator port but readiness probe is collected from application main port due to reliability issues (escaping restarts of apps under overloading).

Our typical configuration looks like (see `additional-path`):
```yaml
management:
  server:
    port: 8081
  endpoint:
    health:
      probes.enabled: true
      group:
        liveness:
          include: livenessState
          additional-path: server:/livez
        readiness:
          include: readinessState
          additional-path: server:/readyz
```

It would be nice to automatically disable observations for additional endpoints as well (according to `additional-path`).

**Comment by izeye** [comment]
Based on https://github.com/spring-projects/spring-boot/pull/36455#issuecomment-2859132042, it seems that the target milestone needs to be updated to the "4.x".

**Comment by wilkinsona** [comment]
This is blocked by https://github.com/micrometer-metrics/micrometer/issues/4061.
